### PR TITLE
Remove introspection endpoint and add MTLS role checks

### LINKED ISSUE #43: Add MTLS role checks
We have MTLS role checking implemented in https://github.com/icebreakerone/demo-energy-mtlsonly/blob/dbf8b841e8b741fb4c076f77195718813ead5944/resource/api/main.py#L72-L85, we should use that here too. We don't need the Perseus directory for that part, we can just update the script for creating certificates to include the role name.

### LINKED ISSUE #44: JWT introspection with authentication app is unnecessary
MTLS with Permissions will use very short lived OAuth access tokens, potentially with less than 1 hour validity. The refresh token will be much longer lived. As such, validation of the JWT just requires it has a valid signature and hasn't expired. 

The introspection endpoint in the authentication app and use in the resource app becomes unnecessary, and should be removed.

### PR DESCRIPTION
**Pull request by kipparker**
Fixes #44 
Fixes #43 

- remove introspection endpoint
- implement signature and certificate binding checks in resource api

### CODE REVIEW

File: resource/api/auth.py [hunk 1]
```diff
@@ -23,6 +30,65 @@ class AccessTokenInactiveError(AccessTokenValidatorError):
...
+
+    If a certificate is present, on our deployment it will be in request.headers['X-Amzn-Mtls-Clientcert']
+    nb. the method and naming of passing the client certificate may vary depending on the deployment
+    """
+    cert_data = unquote(client_certificate).encode("utf-8")
+    cert = x509.load_pem_x509_certificate(cert_data)
+    return cert
+
+
+def _check_certificate(cert: x509.Certificate, decoded_token: dict):
```
Comment (kipparker) [inline resource/api/auth.py:57]: We check the thumbprint in the jwt against the presented client certificate

File: resource/api/auth.py [hunk 2]
```diff
@@ -35,65 +101,47 @@ def get_openid_configuration(issuer_url: str) -> dict:
...
-        raise AccessTokenValidatorError("Introspection request failed")
-    return response.json()
-
-
-def introspect(
-    client_certificate: str, token: str, x_fapi_interaction_id: Optional[str] = None
+def check_token(
+    client_certificate: str | None,
+    token: str,
+    x_fapi_interaction_id: Optional[str] = None,
```
Comment (kipparker) [inline resource/api/auth.py:107]: Follow the Fapi spec on headers - if an interaction is provided, use that, otherwise generate and add one to the response

File: resource/api/auth.py [hunk 3]
```diff
@@ -35,65 +101,47 @@ def get_openid_configuration(issuer_url: str) -> dict:
...
-        or introspection_response["active"] is not True
-    ):
-        raise AccessTokenInactiveError(
-            "Invalid introspection response, does not contain 'active' or is not True"
-        )
+    cert = x509.load_pem_x509_certificate(bytes(unquote(client_certificate), "utf-8"))
+    client_id = cert.subject.get_attributes_for_oid(x509.NameOID.COMMON_NAME)[0].value
+    # get the jwks endpoints from well-known configuration
+    openid_config = get_openid_configuration(conf.ISSUER_URL)
+    jwks_uri = openid_config["jwks_uri"]
```
Comment (kipparker) [inline resource/api/auth.py:126]: Use jwks to check the jwt signature

File: resource/api/auth.py [hunk 4]
```diff
@@ -35,65 +101,47 @@ def get_openid_configuration(issuer_url: str) -> dict:
...
+    key = jwks_client.get_signing_key(header["kid"]).key
+    decoded = jwt.decode(token, key, [header["alg"]], audience=client_id)
+    # Examples of tests to apply
+    if decoded["aud"] != client_id:
+        raise AccessTokenAudienceError("Invalid audience")
+    if decoded["exp"] < int(time.time()):
+        raise AccessTokenTimeError("Token expired")
+    if decoded["iat"] > int(time.time()):
+        raise AccessTokenTimeError("Token issued in the future")
+    _check_certificate(cert, decoded)
```
Comment (kipparker) [inline resource/api/auth.py:141]: There are four checks here to say that the token is valid, do we need to add any more?

**Review by kipparker — COMMENTED** [review]
@bensummers Introspection endpoint removed, and token and certificate checking moved to the resource api

File: resource/api/auth.py [hunk 5]
```diff
@@ -151,3 +152,16 @@ def check_token(
         log.debug(f"using existing interaction ID = {x_fapi_interaction_id}")
     headers["x-fapi-interaction-id"] = x_fapi_interaction_id
     return decoded, headers
+
+
+def require_role(role_name, quoted_certificate) -> bool:
```
Comment (bensummers) [inline resource/api/auth.py:157]: certificate_has_role might be a better name as it doesn't throw?

File: scripts/certmaker.sh [hunk 6]
```diff
@@ -1,35 +1,60 @@
-# Perseus Trust Framework CA 
+# Certificate trees generated:
```
Comment (kipparker) [inline scripts/certmaker.sh:1]: Update scripts to include new role encoding

**Review by kipparker — COMMENTED** [review]
@bensummers new style roles have been added here now. There's a lot of duplication between the two apps (and the mtls only repo), after merging and testing these changes I'll look at moving the duplicate code into a library. For now I'm keen to get this merged into preprod before this pull request gets too large

File: authentication/api/certificate_extensions.py [hunk 7]
```diff
@@ -0,0 +1,20 @@
+from asn1crypto.core import SequenceOf, UTF8String
+
+
+class Roles(SequenceOf):
+    _child_spec = UTF8String
+
+
+def decode(der_bytes: bytes):
```
Comment (bensummers) [inline authentication/api/certificate_extensions.py:8]: Slight quibble with the names here. `decode_roles()` would be better for this function, because there is also a Application extension.

Similar for `encode()`

File: authentication/api/main.py [hunk 8]
```diff
@@ -189,6 +150,16 @@ async def token(
     """
     if x_amzn_mtls_clientcert is None:
         raise HTTPException(status_code=401, detail="No client certificate provided")
+    try:
+        auth.require_role(
+            "https://registry.core.ib1.org/scheme/perseus/role/carbon-accounting",
+            x_amzn_mtls_clientcert,
+        )
+    except auth.CertificateError as e:
```
Comment (bensummers) [inline authentication/api/main.py:158]: When I wrote the original of this, I was hoping it could just be a simple statement. Catching and raising another feels ugly. Rename to `require_role_for_request()` and just throw the `HTTPException`?

File: resource/api/auth.py [hunk 9]
```diff
@@ -35,65 +90,49 @@ def get_openid_configuration(issuer_url: str) -> dict:
...
+    cert = parse_cert(client_certificate)
+    client_id = str(
+        cert.subject.get_attributes_for_oid(x509.NameOID.COMMON_NAME)[0].value
+    )
+    # get the jwks endpoints from well-known configuration
+    openid_config = get_openid_configuration(conf.ISSUER_URL)
+    jwks_uri = openid_config["jwks_uri"]
+    jwks_client = jwt.PyJWKClient(jwks_uri)
+    ssl._create_default_https_context = (
+        ssl._create_unverified_context
```
Comment (bensummers) [inline resource/api/auth.py:120]: Why is this necessary? We shouldn't have something like this in code which might be copied and pasted.

File: resource/api/auth.py [hunk 10]
```diff
@@ -35,65 +90,49 @@ def get_openid_configuration(issuer_url: str) -> dict:
...
+    )  # Must be removed for production
+    header = jwt.get_unverified_header(token)
+    key = jwks_client.get_signing_key(header["kid"]).key
+    decoded = jwt.decode(token, key, [header["alg"]], audience=client_id)
+    # Examples of tests to apply
+    if decoded["aud"] != client_id:
```
Comment (bensummers) [inline resource/api/auth.py:126]: `decoded["client_id"]` not `"aud"

File: resource/api/auth.py (hunk 4 above)
Comment (bensummers) [reply resource/api/auth.py:132]: Check that `"aud"` is the identifier of this data service -- that would be the URL of the Catalog Entry, which is assigned by the publisher in the metadata they generate.

File: resource/tests/__init__.py [hunk 11]
```diff
@@ -1,35 +1,83 @@
...
-    else:
-        introspection_response["exp"] = time.time() + 20
-    if iat:
-        introspection_response["iat"] = iat
-    else:
-        introspection_response["iat"] = time.time() - 20
-    introspection_response["active"] = active
-    return introspection_response
+    # Calculate the thumbprint of the certificate
+    cert_thumprint = (
```
Comment (bensummers) [inline resource/tests/__init__.py:78]: spelling!

File: resource/tests/test_auth.py [hunk 12]
```diff
@@ -15,20 +17,34 @@ def mock_check_token(mocker):
...
+@patch("api.auth.get_openid_configuration")
+@patch("api.auth.jwt.PyJWKClient")
+def test_check_token_integration(mock_jwk_client, mock_get_openid_config):  # noqa
+
+    cert_pem, private_key_pem, private_key, cert_thumbprint = (
+        client_certificate()
+    )  # noqa
+    headers = {"alg": "RS256", "kid": "testkey"}
+    payload = {
+        "aud": CLIENT_ID,
```
Comment (bensummers) [inline resource/tests/test_auth.py:29]: URL of this data service, put CLIENT_ID in `client_id`

**Review by bensummers — CHANGES_REQUESTED** [review]
JWTs need `client_id` and correct use of `aud`.

Few minor typos, and some fussing over names.

The certificate handling is spread across many files, when it's turned into a library it should be encapsulated in a single class, but I think this works for now.

File: authentication/api/main.py (hunk 8 above)
Comment (kipparker) [reply authentication/api/main.py:158]: Starting to think of its use as a library, where the function could be used in different contexts - the library throws its own descriptive errors, you can deal with it as needed - in this case we're catching any kind of CertificateError and returning an http error

**Comment by kipparker** [comment]
> The certificate handling is spread across many files, when it's turned into a library it should be encapsulated in a single class, but I think this works for now.

Possibly, we'll see when we get to it. I prefer functions to classes unless they're needed, makes for simpler code

**Comment by bensummers** [comment]
> > The certificate handling is spread across many files, when it's turned into a library it should be encapsulated in a single class, but I think this works for now.
> 
> Possibly, we'll see when we get to it. I prefer functions to classes unless they're needed, makes for simpler code

The code has evolved over time, and certificates are decoded in auth.py, extension ASN.1 structures encoding in certificate_extensions, OID constants repeated in a few files, and the tests construct certs with extensions. I just meant it should be brought together in a single python module?
